app/controllers/incomeSources/cease/CeaseCheckIncomeSourceDetailsController.scala

```diff
@@ -0,0 +1,242 @@
+              viewModel = viewModel,
+              isAgent = isAgent,
+              backUrl = routes.CeaseIncomeSourceController.show().url,
+              messagesPrefix = messagesPrefix)))
+          case Left(ex) =>
+            Future.failed(ex)
+        }
+      case (Right(None), Right(Some(cessationEndDate))) =>
+        incomeSourceDetailsService.getCheckCeasePropertyIncomeSourceDetailsViewModel(sources, cessationEndDate, incomeSourceType) match {
+          case Right(viewModel) =>
+            Future.successful(Ok(view(
+              viewModel = viewModel,
+              isAgent = isAgent,
+              backUrl = routes.CeaseIncomeSourceController.show().url,
+              messagesPrefix = messagesPrefix)))
+          case Left(ex) =>
+            Future.failed(ex)
+        }
+      case (Right(None), Right(None)) =>
+        val errorMessage = "Both incomeSourceId and cessationEndDate are missing from the session."
+        Future.failed(new Exception(errorMessage))
+      case (Right(Some(_)), Right(None)) =>
+        val errorMessage = s"CessationEndDate is missing from the session."
+        Future.failed(new Exception(errorMessage))
+      case (Left(ex), _) =>
+        val errorMessage = s"Could not get incomeSourceId from session incomeSourceType = $incomeSourceType, ${ex.getMessage}"
+        Future.failed(new Exception(errorMessage))
+      case (_, Left(ex)) =>
+        val errorMessage = s"Could not get cessation date from session incomeSourceType = $incomeSourceType, ${ex.getMessage}"
+        Future.failed(new Exception(errorMessage))
+    }
+
+  } recover {
+    case ex: Exception =>
+      Logger("application").error(s"[CeaseCheckIncomeSourceDetailsController][handleRequest]${if (isAgent) "[Agent] "}" +
+        s"Error getting CeaseCheckIncomeSourceDetails page: ${ex.getMessage}")
+      Redirect(controllers.incomeSources.cease.routes.IncomeSourceNotCeasedController.show(isAgent, SelfEmployment))
+  }
+
+
+  def show(incomeSourceType: IncomeSourceType): Action[AnyContent] =
+    (checkSessionTimeout andThen authenticate andThen retrieveNino andThen retrieveIncomeSources andThen retrieveBtaNavBar).async {
+      implicit user =>
+        handleRequest(
+          sources = user.incomeSources,
+          isAgent = false,
+          None,
+          incomeSourceType = incomeSourceType
+        )
+    }
+
+  def showAgent(incomeSourceType: IncomeSourceType): Action[AnyContent] = Authenticated.async {
+    implicit request =>
+      implicit user =>
+        getMtdItUserWithIncomeSources(incomeSourceDetailsService).flatMap {
+          implicit mtdItUser =>
+            handleRequest(
+              sources = mtdItUser.incomeSources,
+              isAgent = true,
+              incomeSourceType = incomeSourceType
+            )
+        }
+  }
+
+  def handleSubmitRequest(isAgent: Boolean, incomeSourceType: IncomeSourceType)(implicit user: MtdItUser[_], request: Request[_]): Future[Result] = withIncomeSourcesFS {
```

Comment:
Could fix:
This function could be split into smaller ones for readability

Reply:
Maybe one submit function which calls other functions as and when required e.g. an updateCessationDate function?
I've just realised the cyclomatic complexity of this function is 28 which is very high, it'd be good if you could reduce this somehow

Reply:
WIP could pick up when you're free